### PR TITLE
test(typescript): add typescript test

### PR DESCRIPTION
**Pull request by azu**
Add TypeScript test for definition typed.

### CODE REVIEW

File: test/typescript/almin-test.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        this.dispatch({
+            type: "ChildUseCase",
+            value
+        });
+    }
+}
+class ParentUseCase extends UseCase {
+    execute(value: string) {
+        // TODO: improve `execute` signature
+        return this.context.useCase(new ChildUseCase()).execute(value);
```
Comment (azu) [inline test/typescript/almin-test.ts]: Currently, `this.context.useCase(new ChildUseCase()).execute(1);` is pass.
We want to improve this.

File: test/typescript/almin-test.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+}
+const parentUseCase = new ParentUseCase();
+// UseCase - execute
+context.useCase(parentUseCase).execute("value").then(() => {
+    // TODO: Can improve without StoreState?
+    const state = context.getState<StoreState>();
```
Comment (azu) [inline test/typescript/almin-test.ts]: Currently, We should `StoreState` interface.
Could we get it more easily?

Imagination code

```ts
type StoreState = createUnionState(AStore, BStore);
```

or

```ts
context.getState<AStore, BStore>();
```

File: package.json [hunk 3]
```diff
@@ -29,11 +29,11 @@
...
     "build": "npm-run-all -s clean build:src build:lib",
     "build:src": "npm-run-all -p build:src:*",
-    "build:src:cp_obj": "cpx 'src/**/*.{js, jsx}' __obj/ --preserve",
+    "build:src:cp_obj": "cpx 'src/**/*.{js, jsx}' __obj/src --preserve",
     "build:src:tsc": "tsc --project .",
     "build:lib": "npm-run-all -p build:lib:*",
-    "build:lib:cp_lib": "cpx '__obj/**/*.d.ts' lib/ --preserve",
-    "build:lib:down_transform": "cross-env NODE_ENV=production babel __obj/ --out-dir lib/ --source-maps",
+    "build:lib:cp_lib": "cpx '__obj/src/**/*.d.ts' lib/ --preserve",
+    "build:lib:down_transform": "cross-env NODE_ENV=production babel __obj/src --out-dir lib/ --source-maps",
```
Comment (azu) [inline package.json]: tsc create files to  `__obje/{src, test}/`.

File: src/index.ts [hunk 4]
```diff
@@ -10,4 +10,5 @@ export { CompletedPayload } from "./payload/CompletedPayload";
 export { DidExecutedPayload } from "./payload/DidExecutedPayload";
 export { Payload } from "./payload/Payload";
 export { ErrorPayload } from "./payload/ErrorPayload";
-export { WillExecutedPayload } from "./payload/WillExecutedPayload";
\ No newline at end of file
+export { WillExecutedPayload } from "./payload/WillExecutedPayload";
+export { DispatcherPayloadMeta } from "./DispatcherPayloadMeta";
```
Comment (azu) [inline src/index.ts]: I've exports `DispatcherPayloadMeta` for typing.
Could we only export `DispatcherPayloadMeta` as a Type?

I concern that exporting actual `DispatcherPayloadMeta` and the `isTruest` is not actual trusted property.

File: test/typescript/almin-test.ts (hunk 2 above)
Comment (saneyuki) [inline test/typescript/almin-test.ts]: > Could we get it more easily?

Umm, this is the one of weakpoints of the current TypeScript.  Fundamentally, we need to wait [Variadic Type Proposal for TypeScript](https://github.com/Microsoft/TypeScript/issues/5453) to resolve this problem with straight way.

As other approach, we can use multiple overloading. This technique is used in rxjs' `zip`, `combineLatest`, and `Object.assign()`.

```typescript
    getState<T1, T2>(): T1 & T2;
    getState<T1, T2, T3>(): T1 & T2 & T3;
    getState<T1, T2, T3, T4>(): T1 & T2 & T3 & T4;
    getState<T1, T2, T3, T4, T5>(): T1 & T2 & T3 & T4 & T5;
    getState<T1, T2, T3, T4, T5, T6>(): T1 & T2 & T3 & T4 & T5 & T6;
    getState<T>(): T {
        // actual implementation.
    }
```

And we can add type param `T` to `StoreLike` like `StoreLike<T>`. Then we can make `StoreLike.getState()` to `getState(): T`. I feel this approach is more straight way than overloading. But we also need to consider if we bind some `StoreLike` into an other container object.

File: src/index.ts (hunk 4 above)
Comment (azu) [inline src/index.ts]: > Will user construct this as a class?
No.
> I think that it's would be better to export only an interface like:
Thanks. We do apply this pattern to `DispatcherPayloadMeta`.

File: test/typescript/almin-test.ts (hunk 1 above)
Comment (azu) [inline test/typescript/almin-test.ts]: Ok. I've filed that as issue
https://github.com/almin/almin/issues/107

File: src/index.ts (hunk 4 above)
Comment (azu) [inline src/index.ts]: fixed https://github.com/almin/almin/pull/101/commits/854d7d6ae21fb7059e0dea8c7c1ee4359b75970a

File: tsconfig.json [hunk 5]
```diff
@@ -39,7 +39,8 @@
     },
 
     "include": [
-        "src/**/*"
+        "src/**/*",
+        "test/**/*.ts"
```
Comment (azu) [inline tsconfig.json]: fixed.

File: test/typescript/almin-test.ts (hunk 2 above)
Comment (azu) [inline test/typescript/almin-test.ts]: > we need to wait Variadic Type Proposal for TypeScript to resolve this problem with straight way.

Oh, I see.
We should not do something about this, now.

**Comment by azu** [comment]
Reflect review results.

**Pull request merged**